andersoncd/weighted_lasso.py

```diff
@@ -0,0 +1,394 @@
+    else:
+        alphas = np.sort(alphas)[::-1]
+
+    n_alphas = len(alphas)
+
+    coefs = np.zeros((n_features, n_alphas), order='F', dtype=X.dtype)
+    kkt_maxs = np.zeros(n_alphas)
+
+    if return_n_iter:
+        n_iters = np.zeros(n_alphas, dtype=int)
+
+    norms_X_col = norm(X, axis=0)
+
+    for t in range(n_alphas):
+        alpha = alphas[t]
+        if verbose:
+            to_print = "##### Computing alpha %d/%d" % (t + 1, n_alphas)
+            print("#" * len(to_print))
+            print(to_print)
+            print("#" * len(to_print))
+        if t > 0:
+            w = coefs[:, t - 1].copy()
+            p0 = max(len(np.where(w != 0)[0]), 1)
+        else:
+            if coef_init is not None:
+                w = coef_init.copy()
+                p0 = max((w != 0.).sum(), p0)
+                R = y - X @ w
+            else:
+                w = np.zeros(n_features, dtype=X.dtype)
+                R = y.copy()
+
+        sol = celer_primal(
+            X, y, alpha, w, R, norms_X_col, weights,
+            max_iter=max_iter, max_epochs=max_epochs, p0=p0, tol=tol,
+            verbose=verbose)
+
+        coefs[:, t] = w.copy()
+        kkt_maxs[t] = sol[-1]
+        # TODO sol[0], sol[1], sol[2][-1]
+        if return_n_iter:
+            n_iters[t] = len(sol[1])
+
+    results = alphas, coefs, kkt_maxs
+    if return_n_iter:
+        results += (n_iters,)
+
+    return results
+
+
+# def _kkt_violation(XtR, weights, alpha):
+#     for j in range(n_features):
+#         if weights[j] > 0:
+#             kkt[j] = max(0, np.abs(XtR[j]) / n_samples - alpha * weights[j]))
+#         else:
+#                 kkt_max=max(kkt_max, np.abs(XtR[j]))
+
+
+def celer_primal(
+        X, y, alpha, w, R, norms_X_col, weights, max_iter=50,
+        max_epochs=50_000, p0=10, tol=1e-4, use_acc=False, K=5, verbose=0):
+    n_samples, n_features = X.shape
+    pen = weights > 0
+    unpen = ~pen
+    n_unpen = unpen.sum()
+    print(n_unpen)
+    p0 = max(p0, n_unpen)
+    obj_out = []
+    lc = norms_X_col ** 2
+
+    if use_acc:
+        last_K_w = np.zeros([K + 1, n_features])
+        U = np.zeros([K, n_features])
+
+    for t in range(max_iter):
+        kkt = _kkt_violation(w, X, R, weights, alpha, np.arange(n_features))
+        kkt_max = np.max(kkt)
+        if verbose:
+            print(f"KKT max violation: {kkt_max:.2e}")
+        if kkt_max <= tol:
+            break
+        # 1) select features
+        ws_size = max(p0 + n_unpen,
+                      min(2 * (w != 0).sum() - n_unpen, n_features))
+        kkt[unpen] = np.inf  # always include unpenalized features
+        ws = np.argsort(kkt)[-ws_size:]
+
+        if verbose:
+            print(f'Iteration {t}, {ws_size} feats in subpb.')
+
+        # 2) do iterations on smaller problem
+        for epoch in range(max_epochs):
+            _cd_wlasso(X, w, R, alpha, weights, lc, ws)
+
+            if use_acc:
```

Comment:
If we don't mean to benchmark it, we can remove this parameter too. What's your take on this ?

Reply:
I would expose the params `use_acc` and `K` in `celer_primal`, but not in `celer_primal_path` and `WeightedLasso`, wdyt?

Reply:
sounds good